### PR TITLE
let rustdoc print the crate version into docs

### LINKED ISSUE #24336: Show library version in documentation generated by rustdoc
I recently was using a library whose online documentation was out of date. One of the method signatures had changed and that wasn't reflected in the documentation. I checked the source code of the project to confirm that the signature had changed, but it also would have been nice to see on the documentation page that the version I was using didn't match up to the documentation version.

### PR DESCRIPTION
**Pull request by QuietMisdreavus**
This PR adds a new unstable flag to rustdoc, `--crate-version`, which when present will add a new entry to the sidebar of the root module, printing the given version number:

Screenshot of a test crate, showing "Version 1.3.37" under the crate name

Closes #24336

(The WIP status is because i don't want to merge this until i can get the std docs to use it, which i need help from rustbuild people to make sure i get right.)

### CODE REVIEW

**Comment by QuietMisdreavus** [comment]
Pings for everyone!

@rust-lang/docs Does this look right? Is this a good place for this? Right now it *only* shows up on the root module's page - nothing else will show this. Should we put it everywhere?

@GuillaumeGomez and @steveklabnik specifically (possibly dev-tools more broadly, but for now i'll still with rustdoc peers) - This is a new command-line flag. In dev-tools meeting today the prevailing air was that this was not RFC worthy since it was straightforward enough, but i want to make extra sure that the other rustdoc stakeholders are on board.

@alexcrichton @Mark-Simulacrum (other rustbuild experts?) - How can I get bootstrap to use this flag when rendering docs for std and friends?

**Comment by GuillaumeGomez** [comment]
Seems to be good. What happens with a very long version name? Does it goes on two lines or does it just stick out of the sidebar?

**Comment by QuietMisdreavus** [comment]
Looks like it wraps within the sidebar:

image

...unless you use a long word without a breaking separator, in which case it sticks out from the sidebar:

image

**Comment by GuillaumeGomez** [comment]
This is a (smaaall) bug then. :p

**Comment by steveklabnik** [comment]
This sounds good to me!

**Comment by camsteffen** [comment]
Use these styles to fix the wrapping issue on most modern browsers.
```css
overflow-wrap: break-word;
word-wrap: break-word; /* deprecated */
word-break: break-word; /* Chrome, non-standard */
```

**Comment by QuietMisdreavus** [comment]
Wow, thanks! That looks much better!

image

I've force-pushed to include that CSS into the update.

**Comment by GuillaumeGomez** [comment]
Now it's all good. Thanks!

@bors: r+ rollup

**Comment by GuillaumeGomez** [comment]
Ah damn, well, r=me once done.

**Comment by QuietMisdreavus** [comment]
Yeah, the rustdoc piece is basically done, but i'd like to see if i can get rustbuild to use this when generating the docs for std and friends before calling it done. I'll ask around again.

**Comment by Mark-Simulacrum** [comment]
You'll need to edit `src/bootstrap/builder.rs` and add `RUSTDOC_CRATE_VERSION` to the `rustdoc_cmd` environment variables. The value can be pulled from `channel::CFG_RELEASE_NUM` (`1.22.0`) or if you want the SHA, `self.build.rust_info.version(...)` will work.

Let me know if you'd like more detailed instructions, or I can probably get you a commit implementing this within a few days.

**Comment by QuietMisdreavus** [comment]
@Mark-Simulacrum That got me most of the way there. It turns out that the docs weren't using `rustdoc_cmd` at all, instead going through cargo. I wound up adding `RUSTDOCFLAGS` to each doc invocation to get the crate version in, but that can introduce problems in compiler-docs mode, since external crates will also get that version number. I also don't know how this will interact if/when Cargo starts providing this flag itself, but for the moment this does what i want:

image

If you think this is fine, then i'm ready to call this good.

**Comment by Mark-Simulacrum** [comment]
Hm, I'm suspecting that `RUSTDOCFLAGS` may not apply correctly with compiler docs, since we want the version to display on more than just the last crate. I think I'd move to unconditionally setting something like `RUSTDOC_CRATE_VERSION` in `cargo` (`src/bootstrap/builder.rs`) and then using that in `src/bin/rustdoc.rs` to set the unstable option.

**Comment by QuietMisdreavus** [comment]
Ooh, i like that option. I've force-pushed with an update that uses that instead.

File: src/bootstrap/builder.rs [hunk 1]
```diff
@@ -574,6 +574,9 @@ impl<'a> Builder<'a> {
             cargo.env("RUSTC_SAVE_ANALYSIS", "api".to_string());
         }
 
+        // For `cargo doc` invocations, make rustdoc print the Rust version into the docs
+        cargo.env("RUSTDOC_CRATE_VERSION", self.build.rust_version());
```
Comment (Mark-Simulacrum) [inline src/bootstrap/builder.rs]: Let's also add this to `rustdoc_cmd` above to make certain it's passed in all cases.

**Review by Mark-Simulacrum — COMMENTED** [review]
I'm happy with the rustbuild changes after this small nit.

File: src/bootstrap/builder.rs (hunk 1 above)
Comment (QuietMisdreavus) [reply src/bootstrap/builder.rs]: Force-pushed with this update.

**Pull request merged**